src/device_trezor/trezor/transport.cpp

```diff
@@ -223,6 +223,30 @@ namespace trezor{
     msg = msg_wrap;
   }
 
+  Transport::Transport(): m_open_counter(0), m_session_counter(0) {
+
+  }
+
+  void Transport::session_begin(){
+    if (m_session_counter <= 0){
+      open();
+      m_session_counter = 1;
+    } else {
+      m_session_counter += 1;
+    }
+  }
+
+  void Transport::session_end(){
+    m_session_counter -= 1;
+
+    if (m_session_counter < 0){
+      MDEBUG("Session already closed. ctr: " << m_session_counter);
+
+    } else if (m_session_counter == 0) {
+      close();
+    }
+  }
```

Comment:
Why am I feeling that there's some amount of duplicate logic in all this :/

Reply:
I've fixed that by moving the open/close counting to `pre_open` and `pre_close` methods.